File: icare-backend/models/UserHealthSettings.js
const mongoose = require('mongoose');

const userHealthSettingsSchema = new mongoose.Schema(
  {
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
      unique: true,
      index: true,
    },

    // Health Mode Settings
    healthModeEnabled: {
      type: Boolean,
      default: false,
    },
    selectedConditions: {
      type: [String],
      enum: ['Diabetes', 'Hypertension', 'Heart Disease', 'Weight Management', 'General'],
      default: ['General'],
    },

    // Tracker Settings - What to track
    trackedVitals: {
      bloodPressure: { type: Boolean, default: true },
      bloodSugar: { type: Boolean, default: true },
      weight: { type: Boolean, default: false },
      water: { type: Boolean, default: true },
      medication: { type: Boolean, default: true },
      steps: { type: Boolean, default: false },
      sleep: { type: Boolean, default: false },
      heartRate: { type: Boolean, default: true },
      temperature: { type: Boolean, default: false },
      oxygenLevel: { type: Boolean, default: false },
    },

    // Daily Goals
    dailyGoals: {
      water: { type: Number, default: 8 }, // glasses
      steps: { type: Number, default: 10000 },
      sleep: { type: Number, default: 8 }, // hours
    },

    // Unit Preferences
    unitPreferences: {
      weight: {
        type: String,
        enum: ['kg', 'lbs'],
        default: 'kg',
      },
      bloodSugar: {
        type: String,
        enum: ['mg/dL', 'mmol/L'],
        default: 'mg/dL',
      },
    },

    // Reminders
    reminders: {
      medication: [
        {
          time: String, // "09:00", "14:00", "21:00"
          enabled: { type: Boolean, default: true },
          label: String, // Optional label like "Morning dose"
        },
      ],
      water: {
        interval: { type: Number, default: 120 }, // minutes
        enabled: { type: Boolean, default: false },
      },
      healthCheck: {
        time: { type: String, default: '09:00' },
        enabled: { type: Boolean, default: false },
      },
    },

    // Consultation Preferences
    consultationPreferences: {
      preferredLanguage: {
        type: String,
        enum: ['English', 'Urdu'],
        default: 'English',
      },
      preferredDoctorGender: {
        type: String,
        enum: ['Male', 'Female', 'No Preference'],
        default: 'No Preference',
      },
      allowHistoryAccess: { type: Boolean, default: true },
      videoQuality: {
        type: String,
        enum: ['Auto', 'High', 'Medium', 'Low'],
        default: 'Auto',
      },
    },

    // Pharmacy Preferences
    pharmacyPreferences: {
      preferredPharmacyId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Pharmacy',
        default: null,
      },
      defaultDeliveryAddress: {
        street: String,
        city: String,
        postalCode: String,
        phone: String,
      },
      deliveryPreference: {
        type: String,
        enum: ['Home Delivery', 'Self Pickup'],
        default: 'Home Delivery',
      },
    },

    // Lab/Diagnostics Preferences
    labPreferences: {
      homeSampleCollection: { type: Boolean, default: true },
      reportDeliveryMethod: {
        type: String,
        enum: ['Email', 'SMS', 'App Only', 'All'],
        default: 'All',
      },
    },

    // Learning/LMS Preferences
    learningPreferences: {
      notifyNewCourses: { type: Boolean, default: true },
      notifyAssignments: { type: Boolean, default: true },
      notifyCertificates: { type: Boolean, default: true },
    },

    // Notification Preferences
    notificationPreferences: {
      appointments: { type: Boolean, default: true },
      prescriptions: { type: Boolean, default: true },
      labResults: { type: Boolean, default: true },
      promotions: { type: Boolean, default: false },
      healthTips: { type: Boolean, default: true },
    },
  },
  {
    timestamps: true,
  }
);

// Method to get vitals to display based on health mode
userHealthSettingsSchema.methods.getRelevantVitals = function () {
  if (!this.healthModeEnabled) {
    // Return all tracked vitals
    return Object.keys(this.trackedVitals).filter((key) => this.trackedVitals[key]);
  }

  // Map conditions to relevant vitals
  const conditionVitalsMap = {
    Diabetes: ['bloodSugar', 'weight', 'medication', 'steps'],
    Hypertension: ['bloodPressure', 'heartRate', 'medication', 'weight'],
    'Heart Disease': ['heartRate', 'bloodPressure', 'steps', 'weight', 'oxygenLevel'],
    'Weight Management': ['weight', 'steps', 'water', 'sleep'],
    General: Object.keys(this.trackedVitals),
  };

  // Combine vitals from all selected conditions
  const relevantVitals = new Set();
  this.selectedConditions.forEach((condition) => {
    const vitals = conditionVitalsMap[condition] || [];
    vitals.forEach((vital) => {
      if (this.trackedVitals[vital]) {
        relevantVitals.add(vital);
      }
    });
  });

  return Array.from(relevantVitals);
};

// Static method to get or create settings for a user
userHealthSettingsSchema.statics.getOrCreate = async function (userId) {
  let settings = await this.findOne({ userId });
  if (!settings) {
    settings = await this.create({ userId });
  }
  return settings;
};

module.exports = mongoose.model('UserHealthSettings', userHealthSettingsSchema);

File: lib/models/user_health_settings.dart
class UserHealthSettings {
  final String? id;
  final String userId;
  final bool healthModeEnabled;
  final List<String> selectedConditions;
  final TrackedVitals trackedVitals;
  final DailyGoals dailyGoals;
  final UnitPreferences unitPreferences;
  final Reminders reminders;
  final ConsultationPreferences consultationPreferences;
  final PharmacyPreferences pharmacyPreferences;
  final LabPreferences labPreferences;
  final LearningPreferences learningPreferences;
  final NotificationPreferences notificationPreferences;

  UserHealthSettings({
    this.id,
    required this.userId,
    this.healthModeEnabled = false,
    this.selectedConditions = const ['General'],
    required this.trackedVitals,
    required this.dailyGoals,
    required this.unitPreferences,
    required this.reminders,
    required this.consultationPreferences,
    required this.pharmacyPreferences,
    required this.labPreferences,
    required this.learningPreferences,
    required this.notificationPreferences,
  });

  factory UserHealthSettings.fromJson(Map<String, dynamic> json) {
    return UserHealthSettings(
      id: json['_id'],
      userId: json['userId'],
      healthModeEnabled: json['healthModeEnabled'] ?? false,
      selectedConditions: List<String>.from(json['selectedConditions'] ?? ['General']),
      trackedVitals: TrackedVitals.fromJson(json['trackedVitals'] ?? {}),
      dailyGoals: DailyGoals.fromJson(json['dailyGoals'] ?? {}),
      unitPreferences: UnitPreferences.fromJson(json['unitPreferences'] ?? {}),
      reminders: Reminders.fromJson(json['reminders'] ?? {}),
      consultationPreferences: ConsultationPreferences.fromJson(json['consultationPreferences'] ?? {}),
      pharmacyPreferences: PharmacyPreferences.fromJson(json['pharmacyPreferences'] ?? {}),
      labPreferences: LabPreferences.fromJson(json['labPreferences'] ?? {}),
      learningPreferences: LearningPreferences.fromJson(json['learningPreferences'] ?? {}),
      notificationPreferences: NotificationPreferences.fromJson(json['notificationPreferences'] ?? {}),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) '_id': id,
      'userId': userId,
      'healthModeEnabled': healthModeEnabled,
      'selectedConditions': selectedConditions,
      'trackedVitals': trackedVitals.toJson(),
      'dailyGoals': dailyGoals.toJson(),
      'unitPreferences': unitPreferences.toJson(),
      'reminders': reminders.toJson(),
      'consultationPreferences': consultationPreferences.toJson(),
      'pharmacyPreferences': pharmacyPreferences.toJson(),
      'labPreferences': labPreferences.toJson(),
      'learningPreferences': learningPreferences.toJson(),
      'notificationPreferences': notificationPreferences.toJson(),
    };
  }
}

class TrackedVitals {
  final bool bloodPressure;
  final bool bloodSugar;
  final bool weight;
  final bool water;
  final bool medication;
  final bool steps;
  final bool sleep;
  final bool heartRate;
  final bool temperature;
  final bool oxygenLevel;

  TrackedVitals({
    this.bloodPressure = true,
    this.bloodSugar = true,
    this.weight = false,
    this.water = true,
    this.medication = true,
    this.steps = false,
    this.sleep = false,
    this.heartRate = true,
    this.temperature = false,
    this.oxygenLevel = false,
  });

  factory TrackedVitals.fromJson(Map<String, dynamic> json) {
    return TrackedVitals(
      bloodPressure: json['bloodPressure'] ?? true,
      bloodSugar: json['bloodSugar'] ?? true,
      weight: json['weight'] ?? false,
      water: json['water'] ?? true,
      medication: json['medication'] ?? true,
      steps: json['steps'] ?? false,
      sleep: json['sleep'] ?? false,
      heartRate: json['heartRate'] ?? true,
      temperature: json['temperature'] ?? false,
      oxygenLevel: json['oxygenLevel'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'bloodPressure': bloodPressure,
      'bloodSugar': bloodSugar,
      'weight': weight,
      'water': water,
      'medication': medication,
      'steps': steps,
      'sleep': sleep,
      'heartRate': heartRate,
      'temperature': temperature,
      'oxygenLevel': oxygenLevel,
    };
  }
}

class DailyGoals {
  final int water;
  final int steps;
  final int sleep;

  DailyGoals({
    this.water = 8,
    this.steps = 10000,
    this.sleep = 8,
  });

  factory DailyGoals.fromJson(Map<String, dynamic> json) {
    return DailyGoals(
      water: json['water'] ?? 8,
      steps: json['steps'] ?? 10000,
      sleep: json['sleep'] ?? 8,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'water': water,
      'steps': steps,
      'sleep': sleep,
    };
  }
}

class UnitPreferences {
  final String weight;
  final String bloodSugar;

  UnitPreferences({
    this.weight = 'kg',
    this.bloodSugar = 'mg/dL',
  });

  factory UnitPreferences.fromJson(Map<String, dynamic> json) {
    return UnitPreferences(
      weight: json['weight'] ?? 'kg',
      bloodSugar: json['bloodSugar'] ?? 'mg/dL',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'weight': weight,
      'bloodSugar': bloodSugar,
    };
  }
}

class MedicationReminder {
  final String time;
  final bool enabled;
  final String? label;

  MedicationReminder({
    required this.time,
    this.enabled = true,
    this.label,
  });

  factory MedicationReminder.fromJson(Map<String, dynamic> json) {
    return MedicationReminder(
      time: json['time'],
      enabled: json['enabled'] ?? true,
      label: json['label'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'time': time,
      'enabled': enabled,
      if (label != null) 'label': label,
    };
  }
}

class WaterReminder {
  final int interval;
  final bool enabled;

  WaterReminder({
    this.interval = 120,
    this.enabled = false,
  });

  factory WaterReminder.fromJson(Map<String, dynamic> json) {
    return WaterReminder(
      interval: json['interval'] ?? 120,
      enabled: json['enabled'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'interval': interval,
      'enabled': enabled,
    };
  }
}

class HealthCheckReminder {
  final String time;
  final bool enabled;

  HealthCheckReminder({
    this.time = '09:00',
    this.enabled = false,
  });

  factory HealthCheckReminder.fromJson(Map<String, dynamic> json) {
    return HealthCheckReminder(
      time: json['time'] ?? '09:00',
      enabled: json['enabled'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'time': time,
      'enabled': enabled,
    };
  }
}

class Reminders {
  final List<MedicationReminder> medication;
  final WaterReminder water;
  final HealthCheckReminder healthCheck;

  Reminders({
    this.medication = const [],
    required this.water,
    required this.healthCheck,
  });

  factory Reminders.fromJson(Map<String, dynamic> json) {
    return Reminders(
      medication: (json['medication'] as List?)
          ?.map((e) => MedicationReminder.fromJson(e))
          .toList() ?? [],
      water: WaterReminder.fromJson(json['water'] ?? {}),
      healthCheck: HealthCheckReminder.fromJson(json['healthCheck'] ?? {}),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'medication': medication.map((e) => e.toJson()).toList(),
      'water': water.toJson(),
      'healthCheck': healthCheck.toJson(),
    };
  }
}

class ConsultationPreferences {
  final String preferredLanguage;
  final String preferredDoctorGender;
  final bool allowHistoryAccess;
  final String videoQuality;

  ConsultationPreferences({
    this.preferredLanguage = 'English',
    this.preferredDoctorGender = 'No Preference',
    this.allowHistoryAccess = true,
    this.videoQuality = 'Auto',
  });

  factory ConsultationPreferences.fromJson(Map<String, dynamic> json) {
    return ConsultationPreferences(
      preferredLanguage: json['preferredLanguage'] ?? 'English',
      preferredDoctorGender: json['preferredDoctorGender'] ?? 'No Preference',
      allowHistoryAccess: json['allowHistoryAccess'] ?? true,
      videoQuality: json['videoQuality'] ?? 'Auto',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'preferredLanguage': preferredLanguage,
      'preferredDoctorGender': preferredDoctorGender,
      'allowHistoryAccess': allowHistoryAccess,
      'videoQuality': videoQuality,
    };
  }
}

class DeliveryAddress {
  final String? street;
  final String? city;
  final String? postalCode;
  final String? phone;

  DeliveryAddress({
    this.street,
    this.city,
    this.postalCode,
    this.phone,
  });

  factory DeliveryAddress.fromJson(Map<String, dynamic> json) {
    return DeliveryAddress(
      street: json['street'],
      city: json['city'],
      postalCode: json['postalCode'],
      phone: json['phone'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (street != null) 'street': street,
      if (city != null) 'city': city,
      if (postalCode != null) 'postalCode': postalCode,
      if (phone != null) 'phone': phone,
    };
  }
}

class PharmacyPreferences {
  final String? preferredPharmacyId;
  final DeliveryAddress? defaultDeliveryAddress;
  final String deliveryPreference;

  PharmacyPreferences({
    this.preferredPharmacyId,
    this.defaultDeliveryAddress,
    this.deliveryPreference = 'Home Delivery',
  });

  factory PharmacyPreferences.fromJson(Map<String, dynamic> json) {
    return PharmacyPreferences(
      preferredPharmacyId: json['preferredPharmacyId'],
      defaultDeliveryAddress: json['defaultDeliveryAddress'] != null
          ? DeliveryAddress.fromJson(json['defaultDeliveryAddress'])
          : null,
      deliveryPreference: json['deliveryPreference'] ?? 'Home Delivery',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (preferredPharmacyId != null) 'preferredPharmacyId': preferredPharmacyId,
      if (defaultDeliveryAddress != null) 'defaultDeliveryAddress': defaultDeliveryAddress!.toJson(),
      'deliveryPreference': deliveryPreference,
    };
  }
}

class LabPreferences {
  final bool homeSampleCollection;
  final String reportDeliveryMethod;

  LabPreferences({
    this.homeSampleCollection = true,
    this.reportDeliveryMethod = 'All',
  });

  factory LabPreferences.fromJson(Map<String, dynamic> json) {
    return LabPreferences(
      homeSampleCollection: json['homeSampleCollection'] ?? true,
      reportDeliveryMethod: json['reportDeliveryMethod'] ?? 'All',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'homeSampleCollection': homeSampleCollection,
      'reportDeliveryMethod': reportDeliveryMethod,
    };
  }
}

class LearningPreferences {
  final bool notifyNewCourses;
  final bool notifyAssignments;
  final bool notifyCertificates;

  LearningPreferences({
    this.notifyNewCourses = true,
    this.notifyAssignments = true,
    this.notifyCertificates = true,
  });

  factory LearningPreferences.fromJson(Map<String, dynamic> json) {
    return LearningPreferences(
      notifyNewCourses: json['notifyNewCourses'] ?? true,
      notifyAssignments: json['notifyAssignments'] ?? true,
      notifyCertificates: json['notifyCertificates'] ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'notifyNewCourses': notifyNewCourses,
      'notifyAssignments': notifyAssignments,
      'notifyCertificates': notifyCertificates,
    };
  }
}

class NotificationPreferences {
  final bool appointments;
  final bool prescriptions;
  final bool labResults;
  final bool promotions;
  final bool healthTips;

  NotificationPreferences({
    this.appointments = true,
    this.prescriptions = true,
    this.labResults = true,
    this.promotions = false,
    this.healthTips = true,
  });

  factory NotificationPreferences.fromJson(Map<String, dynamic> json) {
    return NotificationPreferences(
      appointments: json['appointments'] ?? true,
      prescriptions: json['prescriptions'] ?? true,
      labResults: json['labResults'] ?? true,
      promotions: json['promotions'] ?? false,
      healthTips: json['healthTips'] ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'appointments': appointments,
      'prescriptions': prescriptions,
      'labResults': labResults,
      'promotions': promotions,
      'healthTips': healthTips,
    };
  }
}
